Magento to Dynamics365 - Customers

jannejjj opened the pull request:
closes #94

Checklist for review:
- [ ] Process name is fine, uses normal casing and descriptive
- [ ] Process description is fine, uses normal casing, dots and is descriptive
  - [ ] Element names use normal casing, no unneeded capital letters.
- [ ] Correct casing used for formats and terms, e.g. CSV, FTP, JSON, Avro
  - [ ] Element names do not contain dots
  - [ ] Comments / note element text are in good language and end with dots.
- [ ] Template parameters checked
  - [ ] Parameter descriptions are good and end with dots.
  - [ ] Sensitive parameters are marked as secrets
- [ ] "Turn off logging" switch is on for any task that has sensitive data as output
- [ ] End "balls" like error throws and return statements have titles (double-click) and reasonable information.
  - [ ] Results have good summaries, preferably with statistics of some sort like "processed X files"
  - [ ] Errors have some error info, usually from task (e.g. some ErrorMessage property contains additional error info)
- [ ] Process is organised with groups (dashed boxes)
- [ ] Straight lines are preferred where possible
- [ ] Subprocesses are forbidden
- [ ] Text mode is preferred over Expressions in expressions, e.g. instead of *$"This is {#var.error}"* use *This is {{#var.error}}*
- [ ] Process metadata file is present and content is fine (version number and name)
- [ ] Process export file is present (process.json)
- [ ] Process description file is present (all sections filled in)
- [ ] Process description has a SVG screenshot of the process
- [ ] Tempalte files are going to the correct folder

FrendsSarlinS commented on Templates/E-Commerce/Magento/Magento to Dynamics365 - Customers/long-description.md:
```diff
@@ -0,0 +1,25 @@
+# Implementation and Usage Notes
+
+This template fetches all customers from Magento, and then either inserts or updates them as Contacts in Dynamics365. To determine if a Magento customer already has a corresponding Contact in Dynamics365, the email addresses of the Magento customers are compared to those of the existing Contacts in Dynamics365.
+
```
"Contacts" should be lowercase here, even though it is an object in Dynamics365.

FrendsSarlinS replied:
Also, note the typo which the bot noticed!

FrendsSarlinS reviewed (changes requested):
The process functions great! Review comments:

1. Process variables should follow the CamelCase naming convention even for abbreviations, such as “MagentoBaseUrl”, “Dynamics365TenantId” etc. I asked this from Jefim yesterday and added it to the new pull request template, but it wasn’t added yet. ID and URL and other abbreviations in capital letters are fine in the descriptions, though. No other casing issues here that pokes my eyes.

2. In the loop, the first return has a written error message plus the ErrorMessage from the task, but after the update and insert tasks there seems to be only ErrorMessage. Is the ErrorMessage alone descriptive enough?
3. The C# statement “Map customer data to Dynamics365 format” could be implemented inside the Dynamics365 tasks as Payload.
4. Not major, but about the layout:
    4.1. The connecting arrows seem to be short (the default length, which is fine) in the beginning of the process and grow longer and longer towards the end.
    4.2. Same applies to the throw/return errors’ connectors, which is why the process seems “jumpy” and not consistent to the eye.

jannejjj commented:
> 1. Process variables should follow the CamelCase naming convention even for abbreviations, such as “MagentoBaseUrl”, “Dynamics365TenantId” etc. I asked this from Jefim yesterday and added it to the new pull request template, but it wasn’t added yet. ID and URL and other abbreviations in capital letters are fine in the descriptions, though. No other casing issues here that pokes my eyes.

Changed process variable casing.

> 2. In the loop, the first return has a written error message plus the ErrorMessage from the task, but after the update and insert tasks there seems to be only ErrorMessage. Is the ErrorMessage alone descriptive enough?

Added some more info to the latter returns as well.

> 3. The C# statement “Map customer data to Dynamics365 format” could be implemented inside the Dynamics365 tasks as Payload.

The payload requires some conditional logic based on how many addresses the customer has on Magento, which I don't think can be done straight in the payload, since I was getting an error about no code blocks being allowed when attempting that. 🤔

> 4. Not major, but about the layout:
>    4.1. The connecting arrows seem to be short (the default length, which is fine) in the beginning of the process and grow longer and longer towards the end.
>    4.2. Same applies to the throw/return errors’ connectors, which is why the process seems “jumpy” and not consistent to the eye.

Probably a matter of taste, for my eyes it looks good as long as the process / different branches is vertically aligned, but i made some adjustments anyway, especially to the lengths of the lines in the beginning of the process. 😄

Also fixed the typo our resident rabbit noticed.

FrendsSarlinS commented:
> The payload requires some conditional logic based on how many addresses the customer has on Magento, which I don't think can be done straight in the payload, since I was getting an error about no code blocks being allowed when attempting that. 🤔

Did you try doing it in this format, if the customer has an address:

```
json.address1_line1 = #var.magentoCustomer.addresses.Count > 0 ? #var.magentoCustomer.addresses[0].street[0] : "",
json.address1_country = #var.magentoCustomer.addresses.Count > 0 ? #var.magentoCustomer.addresses[0].country_id : "",
etc
```
and if the customer has more addresses:

```
json.address2_line1 = #var.magentoCustomer.addresses.Count > 1 ? #var.magentoCustomer.addresses[1].street[0] : "",
json.address2_country = #var.magentoCustomer.addresses.Count > 1 ? #var.magentoCustomer.addresses[1].country_id : "",
etc
```

FrendsSarlinS reviewed (approved):
With the new changes, the process looks good to me!

FrendsSarlinS commented on Templates/E-Commerce/Magento/Magento to Dynamics365 - Customers/long-description.md:
```diff
@@ -0,0 +1,25 @@
+This template fetches customers from Adobe Commerce / Magento and inserts or updates them to Dynamics365 as contacts.
+Using this template you can synchronize customers from Adobe Commerce / Magento to Dynamics365.
+
```
Should we add a newline here?
